common/djangoapps/student/models.py

```diff
@@ -3116,6 +3120,130 @@ class AccountRecoveryConfiguration(ConfigurationModel):
     )
 
 
+class UserCelebration(TimeStampedModel):
+    """
+    Keeps track of how we've celebrated a user's progress on the platform.
+    This class is for course agnostic celebrations (not specific to a particular enrollment).
+    CourseEnrollmentCelebration is for celebrations that happen separately for each separate course.
+
+    The first_day_of_streak and last_day_of_streak fields are used to
+    control celebration of the streak feature.
+    A streak is when a learner visits the learning MFE N days in a row.
+    The business logic of streaks for a 3 day streak and 1 day break is the following:
+    1. Each streak should be celebrated exactly once, once the learner has completed the streak.
+    2. If a learner misses enough days to count as a break, the streak resets back to 0.
+    3. The streak is measured against the learner's configured timezone
+    4. We keep track of the total length of the streak, so there is a possibility in the future
+    to add multiple celebrations for longer streaks.
+    5. We keep track of the longest_ever_streak field for potential future use for badging purposes.
+
+    .. no_pii:
+    """
+    user = models.OneToOneField(User, models.CASCADE, related_name='celebration')
+    first_day_of_streak = models.DateField(default=None, null=True, blank=True)
+    last_day_of_streak = models.DateField(default=None, null=True, blank=True)
+    longest_ever_streak = models.IntegerField(default=0)
+    STREAK_LENGTHS_TO_CELEBRATE = [3]
+    STREAK_BREAK_LENGTH = 1
+
+    def __str__(self):
+        return (
+            '[UserCelebration] user: {}; first_day_of_streak {};  last_day_of_streak {}; longest_ever_streak {};'
```

Comment:
if you want to get extra fancy, python3 has format strings so you could actually do something like
`return f'[UserCelebration] user: {self.user.username}; first_day_of_streak {self.first_day_of_streak};  last_day_of_streak {self.last_day_of_streak}; longest_ever_streak {self.longest_ever_streak};'`

Reply:
is this considered better?  
i have rarely seen this way in edx python code

Reply:
That's because it's newly available with python3.8 (I think was a feature in 3.6 or 3.7, but we jumped over those). Better or not is a matter of taste, but I think it's becoming more common.